README.md
# RPi_unattended
Originally started in 2015, this was a fun exploratory project to create a ground-up fault-tolerant (HW &amp; SW) platform for unattended RPi projects. Includes sample use-case of displaying a weather forecast on a PiTFT 2.8" screen

# Background
I've always been fascinated by the ability of satellites, space probes, mars rovers, etc. to be controlled, recovered, and re-programmed completely remotely.
With this project, I set out to understand some of the requirements that these platforms must meet for that to be possible.
It's really amazing what bases need to be covered when you can't just "unplug it and plug it in again" :-)

# Core Features / Objectives
- Operate completely unattended 24/7/365.
- Recover gracefully from any software crashes via watchdog
- Operate in a low-power 'deep sleep' mode for extended periods of main power loss (up to 1 month)
- Resume normal operation when main power is available again
- Recover gracefully from network connection loss
- Provide failsafe keyboard-less & mouse-less full access for on-site service
- Provide onsite physical 'kill switch' for reboot/shutdown
- Automatically notify admin via email of bootup, shutdown, main power lost, and other significant events

# Dependencies
- PiJuice battery hat (originally started with Pimodules' UPSPico)
- pywapi (however this is now obsolete, and a new weather api is needed, perhaps owm...)
- PiTFT 2.8"
- Some other software (mutt, logger, etc.)

# Plans
- Bring project to a more modern platform (originally built on an Rpi2b)
- Clean up code and centralization of user configuration (into a json or yaml file)
- Update the weather applications api to something current.


changelog.txt
RPi_unattended CHANGELOG

TODO LIST:
====================================================================================================================================

CHANGE LOG:
====================================================================================================================================

[Sat 30 Jun 2018]
- commented out bottom lines of /etc/rsyslog.conf due to a bug I found.
- still trying to prevent occasional freezing of weather app...
- moved radar download and ip address check to updateweather(). now it only freezes at :00, so the problem must be in update weather.

[Tue 7 Jun 2018]
- created unplug2reboot
- modified checkpower to set rtc alarm and wakeup schedule when needed, off otherwise.

[Mon 6 Jun 2018]
- Massive overhaul to replace PicoUPS with Pijuice:
-- Updated to Raspbian Jessie
-- Migrated rc.local scripts to systemd services in /etc/systemd/system (rc.local isn't really used anymore)
-- Various updates to weather script to support pijuice and running even in case of no network detection
-- Weather script now shows ip address as green (internet reachable), orange (not reachable), red (error)
-- Updates to unplug2shutdown to support PiJuice
-- checkpower.py script  introduced with PiJuice (when usb power lost, shut down and periodically wake up for a little bit to check)
-- Added RTC support from Pijuice (apparently systemd-timesync updates the time automatically when NTP is available)

[Sat 26 Nov 2017]
- changed batpercentage calculation in ~/projects/weather/weather.systemstatus.py to match firmware f6 values.

[Wed 22 Nov 2017]
- Now turning off lcd backlight on power down when still plugged in. Temporarily done by editing /etc/rc0.d/K10halt script.

[Mon 20 Nov 2017]
- Updated pico firmware to 0xf6, testing pico's 'stay alive' watchdog to help fix 'mysterious freezes/shutdowns'. See stayalive.sh
- Enabled logrotation for /var/log/rclocal.log. See /etc/logrotate.d/rclocal.

[Thu 30 Mar 2017]
- updated pico firmware to 0x31, changed systemstatus.py to 4.4v from 4.26 to reflect new voltage level to get to 100%

[Sun 19 mar 2017]
- Updated Pico Firmware to 0x30, and programmed all leds off (no more blinking UPS led at night :)).
- Minor change to weather.systemstatus.py: script now creates real file and saves it for radar image, then displays it. I have a funny feeling
that the 'mystery freezes' might be due to the radar image. Before, a filestream was created directly from the url, and I don't know what happens
if the url doesn't work. In the new case, it should always display radar.jpg, even if it wasn't updated. We'll see...

[Wed 8 mar 2017]
- created powersaver script to combine all of the on/off scripts in power/powersaving.
- minor edit to spinup wifi to do networking stop/start, ifdown instead of ifconfig wlan0 down, and log everything to syslog.
- symlinked all of my scripts to /home/matt/bin and /usr/local/bin. I might centralize the location of all the scripts I've written...
- added to all bash scripts (exec 1> >(logger -s -t $(basename $0)) 2>&1) to log everything to syslog (wifiseeker no longer uses wifiseeker.log)

[Tuesday 7 Mar 2017]
- cleaned up some scripts & logs. Most notable is powerstatestatus, spinupwifi, and rc.local. Added startup sms message to systemmailer.

[Monday 6 Mar 2017]
- Installed mutt and merged mailers into headless/systemmailer.sh. mailx and uuencode didn't work, since iphone can't read uuencoded attachments.
- systemmailer.sh can take in command line arguments startup, shutdown, powerlost, powerrestored, and custom to send emails.
- Mailer subsystems is as follows:
  Startupmailer: systemmailer startup is run in rc.local.
  Shutdownmailer: init.d/shutdownmailer daemon calls /usr/local/bin/shutdownmailer, which calls systemmailer shutdown.
  Powerlostmailer: powerlostmailer script is run in rc.local, which calls systemmailer powerlost/powerrestorred.

[Sunday 5 mar 2017]
- Changed mailers. Bootup is startupmailer.py in rc.local. Shutdown is /init.d/shutdownmailer (status checked in rc.local). Powerlost is powerlostmailer in rc.local
- Todo: see if this can be cleaned up a little more

[Saturday 4 mar 2017]
- big cleanup of rc.local. Not doing direct commands there anymore. instead loading scripts to do everything. Most notable is adoption of spinupwifi.sh.
- figured out how to run things in the background, but start them with delay. eg: in rc.local, bootup mail starts up before end, but waits 10s before sending to catch it all.

[Monday 27 feb 2017]
- Alertmailer script now calls checkpower.py, dumps the output to a log, and sends the log along with the mail on the shutdown mail

[Sunday 26 feb 2017]
- installed mpack to send attachments via email from commandline.
- installed dataplicity to access rpi over internet (dataplicity.com)

[Saturday 25 Feb 2017]
- big update: created systemstatus.py out of weather.picoups.py. displays pico things, etc. several hours went into this
- wifiseeker now writes status to file (idle or seeking). systemstatus polls file to display status
- removed 'internet access' check from systemstatus.py. pinging google caused big lag in script if offline. not really needed anyway.

[Monday 20 Feb 2017]
- created alertmailer program to send email on main power lost/restored, and shutdown. see /usr/local/bin/alertmailer and /etc/init.d/alertmailer
- created weather.picoups.py (replaces weather.py on startup). Now display system status via pico ups i2c commands.
- weather now starts by rc.local calling startweather.sh, which will redirect any output to ~/log/weather.log
- updated Rc.local the other day to add iphone hotspot
- testing /communications/wifiseeker in rc.local which hopefully will try to reconnect to internet every 10 minutes (via spinupwifi.sh).

[Monday 16 Feb 2017]
- Changed a lot in the last few weeks, might not remember it all.
- Added pico ups hat as backup battery.

icon_defs.py

sd = './icons/'	# Icon sub-directory name.

# The forcast includes an icon index number.  Below is an ordered array that has the
# icon filename for the icon index.
icons = [ 
	'thunderstorms.png',	# 0 - Thunderstorms
	'windy.png',			# 1 - Windy Rain
	'unknown.png',			# 2 - Windy Rain
	'thunderstorms.png',	# 3 - Thunderstorms
	't-storms2.png',		# 4 - T-Storms
	'rain_snow.png',		# 5 - Rain Snow
	'rain_snow.png',		# 6 - Rain Sleet
	'rain_snow.png',		# 7 - Snow/Rain Icy Mix
	'rain_snow.png',		# 8 - Freezing Drizzle
	'rain.png',				# 9 - Drizzle
	'rain.png',				#10 - Freezing Rain
	't-showers.png',			#11 - T-Showers
	#'unknown.png',			#11 - Showers
	#'unknown.png',			#11 - Light Rain
	'rain.png',				#12 - Heavy Rain
	'snow.png',				#13 - Snow Flurries
	'snow.png',				#14 - Light Snow
	'snow.png',				#15 - Snowflakes
	'snow.png',				#16 - Heavy Snow
	'thunderstorms.png',	#17 - Thunderstorms
	'unknown.png',			#18 - Hail
	'unknown.png',			#19 - Dust
	'fog.png',				#20 - Fog
	'fog.png',				#21 - Haze
	'fog.png',				#22 - Smoke
	'windy.png',			#23 - Windy
	'windy.png',			#24 - Windy
	'sunny.png',			#25 - Frigid
	'partly_cloudy.png',	#26 - Cloudy
	'partly_cloudy.png',	#27 - Mostly Cloudy Night (the "Night" will not be included)
	'mostly_cloudy.png',	#28 - Mostly Cloudy
	'partly_cloudy.png',			#29 - Partly Cloudy Night (the "Night" will not be included)
	'partly_cloudy.png',			#30 - Partly Cloudy
	'clear_night.png',		#31 - Clear Night (the "Night" will not be included)
	'sunny.png',			#32 - Sunny
	'sunny.png',			#33 - Fair
	#'unknown.png',		#33 - Mostly Clear Night (the "Night" will not be included)
	'sunny.png',			#34 - Fair
	#'unknown.png',		#34 - Mostly Sunny
	'thunderstorms.png',	#35 - Thunderstorms
	'sunny.png',			#36 - Hot
	'scattered_tstorms.png',#37 - Isolated Thunder
	'scattered_tstorms.png',#38 - Scattered T-Storms
	'rain.png',				#39 - Scattered Rain
	'rain.png',				#40 - Heavy Rain
	'snow.png',				#41 - Scattered Snow
	'snow.png',				#42 - Heavy Snow
	'snow.png',				#43 - Windy/Snowy
	'mostly_cloudy.png',	#44 - Partly Cloudy Day
	'rain.png',				#45 - Scattered Showers Night (the "Night" will not be included)
	'snow.png'	,			#46 - Snowy Night
	'scattered_tstorms.png'	#47 - Scattered T-Storms Night (the "Night" will not be included)
]
